File: teds/lib/surface.py
# This source code is licensed under the 3-clause BSD license found in
# the LICENSE file in the root directory of this project.
"""Tools to simulate surface reflection"""
import numpy as np
import numpy.typing as npt


class Surface:
    """Collection of methods to calculate surface properties"""
    def __init__(self, wave: npt.NDArray[np.float64]) -> None:
        wave_max = np.amax(wave)
        wave_min = np.amin(wave)
        wave_mean = np.mean(wave)
        self.spec = (wave - wave_mean) / (wave_max - wave_min)
        self.alb = np.zeros_like(wave)

    def get_albedo_poly(self, alb_coeff: list[float]) -> None:
        """Get albedo.

        Parameters
        ----------
        spec
            Spectral array (can be any spectral quantity)
        alb
            List of albedo coefficients

        Returns
        -------
            Albedo polynomial dependent on wavelength.

        """
        self.alb[:] = 0
        for i in range(len(alb_coeff)):
            self.alb += alb_coeff[i] * self.spec**i

File: cpp/l1l2/surface.cpp
// This source code is licensed under the 3-clause BSD license found
// in the LICENSE file in the root directory of this project.

#include "surface.h"

#include <algorithm>
#include <numeric>

namespace tango {

Surface::Surface(const int n_albedos, const Eigen::ArrayXd& wave_lbl)
{
    waves.resize(wave_lbl.size(), n_albedos);
    waves.col(0) = (wave_lbl - wave_lbl.mean())
                   / (wave_lbl.maxCoeff() - wave_lbl.minCoeff());
    for (int i { 1 }; i < n_albedos; ++i) {
        waves.col(i) = waves.col(i - 1) * waves.col(0);
    }
    alb.resize(wave_lbl.size());
}

auto Surface::getAlbedoPoly(const Eigen::Ref<const Eigen::ArrayXd> albedo_coeff)
  -> void
{
    alb = albedo_coeff(0);
    for (long int i { 1 }; i < albedo_coeff.size(); ++i) {
        alb += albedo_coeff(i) * waves.col(i - 1);
    }
}

} // namespace tango
